[wasm] Compile .bc->.o in parallel, before passing to the linker

radical opened the pull request:
Currently, we use `mono-aot-cross` to compile the managed assemblies
into bitcode files. Then we pass the `.bc` files to `emcc` for linking.
We can save some amount of work, if precompile `.bc` files to `.o`
files, and pass them for linking.

This would also allow us to rebuild the `.o` files only when needed,
thus improving incremental builds.
Also, emscripten has recently removed some amount of parallelization, so
this allows us to do some pre-linking work in parallel, and speed up the
builds a little bit.

This is implemented via a new `EmccCompile` task, that is meant *only*
for compiling, and not as a general purpose `emcc` invocation task.

Also, this bumps msbuild used for tasks compilation to `16.9.0`.

radical committed:
Utils.RunProcess: some refactoring

- this cleans up `Utils.RunProcess` a bit, but doesn't change the
  behavior for the existing non-wasm users (Apple, and Android build
  tasks)

- New overloads are added to support getting the output, and exit code,
  and throwing only when really needed
- And running shell commands, appropriately on windows, and unix.

radical committed:
[wasm] Compile .bc -> .o files, in parallel

Currently, we use `mono-aot-cross` to compile the managed assemblies
into bitcode files. Then we pass the `.bc` files to `emcc` for linking.
We can save some amount of work, if precompile `.bc` files to `.o`
files, and pass them for linking.

This would also allow us to rebuild the `.o` files only when needed,
thus improving incremental builds.
Also, emscripten has recently removed some amount of parallelization, so
this allows us to do some pre-linking work in parallel, and speed up the
builds a little bit.

This is implemented via a new `EmccCompile` task, that is meant *only*
for compiling, and not as a general purpose `emcc` invocation task.

radical committed:
Disable failing tests

`System.IO.MemoryMappedFiles.Tests` - Issue: https://github.com/dotnet/runtime/issues/54194

radical committed:
Merge remote-tracking branch 'origin/main' into wasm-bc-third

radical committed:
[wasm] EmccCompile task: Use msbuild's new resource management API

- `IBuildEngine9.RequestCores`
- This is available only in 16.10+, so bump msbuild used for building
  the tests

radical committed:
cleanup

radical committed:
Correctly handle non-aot native case

radical committed:
Merge remote-tracking branch 'origin/main' into wasm-bc-third

radical commented:
@akoeplinger Does the msbuild bump to 16.10.0 look fine?

radical committed:
Merge remote-tracking branch 'origin/main' into wasm-bc-third

radical committed:
Merge branch 'main' into wasm-bc-to-o-parallel

radical commented:
/azp run runtime,runtime-staging

radical committed:
Merge remote-tracking branch 'origin/main' into wasm-bc-third

radical committed:
Merge remote-tracking branch 'rf/wasm-bc-to-o-parallel' into wasm-bc-third

lewing commented:
/azp run runtime

radical commented:
The browser builds keep failing with:

```
##[error]src\libraries\Microsoft.NETCore.Platforms\tests\Microsoft.NETCore.Platforms.Tests.csproj(0,0): error NU1201: (NETCORE_ENGINEERING_TELEMETRY=Restore) Project System.Security.AccessControl is not compatible with net472 (.NETFramework,Version=v4.7.2). Project System.Security.AccessControl supports: net6.0 (.NETCoreApp,Version=v6.0)
D:\workspace\_work\1\s\src\libraries\Microsoft.NETCore.Platforms\tests\Microsoft.NETCore.Platforms.Tests.csproj : error NU1201: Project System.Security.AccessControl is not compatible with net472 (.NETFramework,Version=v4.7.2) / browser-wasm. Project System.Security.AccessControl supports: net6.0 (.NETCoreApp,Version=v6.0) [D:\workspace\_work\1\s\Build.proj]
##[error]src\libraries\Microsoft.NETCore.Platforms\tests\Microsoft.NETCore.Platforms.Tests.csproj(0,0): error NU1201: (NETCORE_ENGINEERING_TELEMETRY=Restore) Project System.Security.AccessControl is not compatible with net472 (.NETFramework,Version=v4.7.2) / browser-wasm. Project System.Security.AccessControl supports: net6.0 (.NETCoreApp,Version=v6.0)
```

.. but I don't see this on `main` at all, AFAICS. But with the current state of CI, it's difficult to get a clear picture.

It might be related to the msbuild bump in this PR.

radical commented:
@ViktorHofer Any idea why a msbuild bump to 16.10.0 in this PR might be failing wasm/android/iOS/etc builds? That's the only non-wasm specific change here.

> Some of the builds(browser, android, iOS etc) keep failing with:
> 
> ```
> ##[error]src\libraries\Microsoft.NETCore.Platforms\tests\Microsoft.NETCore.Platforms.Tests.csproj(0,0): error NU1201: (NETCORE_ENGINEERING_TELEMETRY=Restore) Project System.Security.AccessControl is not compatible with net472 (.NETFramework,Version=v4.7.2). Project System.Security.AccessControl supports: net6.0 (.NETCoreApp,Version=v6.0)
> D:\workspace\_work\1\s\src\libraries\Microsoft.NETCore.Platforms\tests\Microsoft.NETCore.Platforms.Tests.csproj : error NU1201: Project System.Security.AccessControl is not compatible with net472 (.NETFramework,Version=v4.7.2) / browser-wasm. Project System.Security.AccessControl supports: net6.0 (.NETCoreApp,Version=v6.0) [D:\workspace\_work\1\s\Build.proj]
> ##[error]src\libraries\Microsoft.NETCore.Platforms\tests\Microsoft.NETCore.Platforms.Tests.csproj(0,0): error NU1201: (NETCORE_ENGINEERING_TELEMETRY=Restore) Project System.Security.AccessControl is not compatible with net472 (.NETFramework,Version=v4.7.2) / browser-wasm. Project System.Security.AccessControl supports: net6.0 (.NETCoreApp,Version=v6.0)
> ```

radekdoulik commented on src/mono/wasm/build/WasmApp.Native.targets:
```diff
@@ -165,18 +166,14 @@
 
       <EmccCompileOptimizationFlag Condition="'$(EmccCompileOptimizationFlag)' == ''">$(_EmccOptimizationFlagDefault)</EmccCompileOptimizationFlag>
-      <EmccLinkOptimizationFlag    Condition="'$(EmccLinkOptimizationFlag)' == ''"   >$(_EmccOptimizationFlagDefault)</EmccLinkOptimizationFlag>
+      <EmccLinkOptimizationFlag    Condition="'$(EmccLinkOptimizationFlag)' == ''"   >-O0</EmccLinkOptimizationFlag>
```
We might also try to use `-Wl,-O0` and `-Wl,-lto-O0` here? (context: https://github.com/emscripten-core/emscripten/issues/14485#issuecomment-865213083)

radical replied:
Sure, we can try all that in a follow up PR. Unless, one of those is definitely better than `-O0`, then we can add it here itself.

radical committed:
Merge remote-tracking branch 'origin/main' into wasm-bc-to-o-parallel

radical committed:
Disable Microsoft.NETCore.Platforms.Tests on wasm, for local builds too

- This is meant to be disabled for CI (it has `IgnoreForCI=true` in the
  csproj)

radical committed:
Merge remote-tracking branch 'origin/main' into wasm-bc-to-o-parallel

radical committed:
Undo 16.10.0 msbuild bump

.. because we keep running into the following on browser, and other
platforms:

```
/__w/1/s/src/libraries/Microsoft.NETCore.Platforms/tests/Microsoft.NETCore.Platforms.Tests.csproj : error NU1201: Project System.Security.AccessControl is not compatible with net472 (.NETFramework,Version=v4.7.2). Project System.Security.AccessControl supports: net6.0 (.NETCoreApp,Version=v6.0) [/__w/1/s/Build.proj]
\#\#[error]src/libraries/Microsoft.NETCore.Platforms/tests/Microsoft.NETCore.Platforms.Tests.csproj(0,0): error NU1201: (NETCORE_ENGINEERING_TELEMETRY=Restore) Project System.Security.AccessControl is not compatible with net472 (.NETFramework,Version=v4.7.2). Project System.Security.AccessControl supports: net6.0 (.NETCoreApp,Version=v6.0)
```

radical committed:
Merge remote-tracking branch 'origin/main' into wasm-bc-to-o-parallel

radical committed:
undo test exclusion change related to the 16.10.0 bump

radical committed:
disable Microsoft.NETCore.Platforms.Tests on wasm

This is already not being sent to CI with `IgnoreForCI=true`.
It fails on CI with:

```
/_/src/Shared/NativeMethodsShared.cs(855,29): error IL1005: Microsoft.Build.Shared.NativeMethodsShared.SystemInformation.get: Error processing method 'Microsoft.Build.Shared.NativeMethodsShared.SystemInformationData.SystemInformationData()' in assembly 'Microsoft.Build.Utilities.Core.dll' [/__w/1/s/src/libraries/Microsoft.NETCore.Platforms/tests/Microsoft.NETCore.Platforms.Tests.csproj]
\##[error]/_/src/Shared/NativeMethodsShared.cs(855,29): error IL1005: (NETCORE_ENGINEERING_TELEMETRY=Build) Microsoft.Build.Shared.NativeMethodsShared.SystemInformation.get: Error processing method 'Microsoft.Build.Shared.NativeMethodsShared.SystemInformationData.SystemInformationData()' in assembly 'Microsoft.Build.Utilities.Core.dll'
  Mono.Linker.LinkerFatalErrorException: /_/src/Shared/NativeMethodsShared.cs(855,29): error IL1005: Microsoft.Build.Shared.NativeMethodsShared.SystemInformation.get: Error processing method 'Microsoft.Build.Shared.NativeMethodsShared.SystemInformationData.SystemInformationData()' in assembly 'Microsoft.Build.Utilities.Core.dll'
   ---> System.NotImplementedException: switch
     at Mono.Linker.Steps.UnreachableBlocksOptimizer.BodyReducer.IsConstantBranch(OpCode opCode, Int32 operand) in illink.dll:token 0x6000770+0x41
     at Mono.Linker.Steps.UnreachableBlocksOptimizer.BodyReducer.RemoveConditions() in illink.dll:token 0x6000768+0x22a
     at Mono.Linker.Steps.UnreachableBlocksOptimizer.BodyReducer.RewriteBody() in illink.dll:token 0x6000766+0xa
     at Mono.Linker.Steps.UnreachableBlocksOptimizer.ProcessStack() in illink.dll:token 0x60004e9+0x102
     at Mono.Linker.Steps.UnreachableBlocksOptimizer.ProcessMethod(MethodDefinition method) in illink.dll:token 0x60004e5+0x44
     at Mono.Linker.Steps.MarkStep.ProcessMethod(MethodDefinition method, DependencyInfo& reason, Scope& scope) in illink.dll:token 0x600040e+0x9b
     at Mono.Linker.Steps.MarkStep.ProcessQueue() in illink.dll:token 0x6000399+0x24
     --- End of inner exception stack trace ---
     at Mono.Linker.Steps.MarkStep.ProcessQueue() in illink.dll:token 0x6000399+0xaa
     at Mono.Linker.Steps.MarkStep.ProcessPrimaryQueue() in illink.dll:token 0x6000396+0xa
     at Mono.Linker.Steps.MarkStep.Process() in illink.dll:token 0x6000393+0x0
     at Mono.Linker.Steps.MarkStep.Process(LinkContext context) in illink.dll:token 0x600038a+0x41
     at Mono.Linker.Pipeline.ProcessStep(LinkContext context, IStep step) in illink.dll:token 0x6000264+0x0
     at Mono.Linker.Pipeline.Process(LinkContext context) in illink.dll:token 0x6000263+0xf
     at Mono.Linker.Driver.Run(ILogger customLogger) in illink.dll:token 0x60000fd+0x20
  Optimizing assemblies for size, which may change the behavior of the app. Be sure to test after publishing. See: https://aka.ms/dotnet-illink
/__w/1/s/.dotnet/sdk/6.0.100-preview.4.21255.9/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.ILLink.targets(80,5): error NETSDK1144: Optimizing assemblies for size failed. Optimization can be disabled by setting the PublishTrimmed property to false. [/__w/1/s/src/libraries/Microsoft.NETCore.Platforms/tests/Microsoft.NETCore.Platforms.Tests.csproj]
\##[error].dotnet/sdk/6.0.100-preview.4.21255.9/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.ILLink.targets(80,5): error NETSDK1144: (NETCORE_ENGINEERING_TELEMETRY=Build) Optimizing assemblies for size failed. Optimization can be disabled by setting the PublishTrimmed property to false.
```

radical commented:
cc @ericstj @ViktorHofer for the change disabling `Microsoft.NETCore.Platforms.Tests` for wasm.

radical committed:
[wasm] Disable System.IO.MemoryMappedFiles.Tests

Issue: https://github.com/dotnet/runtime/issues/54194

radical committed:
Merge remote-tracking branch 'origin/main' into wasm-bc-to-o-parallel

radical commented:
1 test failure for wasm:
```
[04:55:37] info: browser: Console websocket connected.
[04:55:37] info: Arguments: --run,WasmTestRunner.dll,System.IO.Compression.Tests.dll,-notrait,category=IgnoreForCI,-notrait,category=OuterLoop,-notrait,category=failing
[04:55:37] fail: [out of order message from the browser]: http://127.0.0.1:34229/runtime.js 234:26 Uncaught TypeError: Cannot set property 'loaded_cb' of null
[05:10:37] fail: Application has finished with exit code TIMED_OUT but 0 was expected
```

@Daniel-Genkin is already looking at fixing this.
I think we can merge this.

lewing commented:
> @Daniel-Genkin is already looking at fixing this.
> I think we can merge this.

fix should be in https://github.com/dotnet/runtime/pull/54587

ViktorHofer commented:
> @ViktorHofer Any idea why a msbuild bump to 16.10.0 in this PR might be failing wasm/android/iOS/etc builds? That's the only non-wasm specific change here.

That very much sounds like https://github.com/NuGet/Home/issues/10368. Usually you can fix this by adding the transitive dependencies, in this case in the Platforms.Tests.csproj file.

radical committed:
Merge remote-tracking branch 'origin/main' into wasm-bc-to-o-parallel

radical commented:
The failures are all unrelated to this PR.

ViktorHofer commented:
> The failures are all unrelated to this PR.

Usually we link to issues that track the failures according to https://github.com/dotnet/runtime/blob/main/docs/issues-pr-management.md#common-policies. Can you please do so for the issues that you were seeing in the last build?

radical commented:
Failures:

`Libraries Test Run checked coreclr windows x86 Release` - https://github.com/dotnet/runtime/issues/54628
`Libraries Test Run release mono OSX x64 Debug` - https://github.com/dotnet/runtime/issues/54624
`Libraries Test Run release coreclr windows x64 Debug` - https://github.com/dotnet/runtime/issues/54627
`Libraries Test Run release coreclr Linux x64 Debug` - https://github.com/dotnet/runtime/issues/54623

And for `CoreCLR Pri0 Runtime Tests Run windows arm64 checked` - `profiler.gc`, and `JIT.Directed` failed but I'm not able to access the logs for them.
